### ISSUE TITLE
unexpected auto-lagging results & other questions

### ISSUE THREAD

**Issue by manderso7** [guilhemmarchand/trackme#104]
Originally posted this on the answers board, but received no response. You may check this page more frequently.

Working a bunch with the Trackme app and it's showing a lot of promise. I finally got the right MLTK and Python applications installed, hopefully that will help some of the issues I'm having.

When I modify a data source, choosing "auto lagging" comes back with unexpected results. In particular, I have a data source that only ingests M-F, between 9am and 5pm. The auto percentile lag of this data source for 7-30 days at 1-3 seconds. How would I go about getting a longer average lag time for this source?

In addition, how can I tell trackme to not show an alert state on the same sourcetype on a monday morning, since it hasn't gotten any events since the friday before? I don't want to set the lag time as too high as that will interfere w/ monitoring during weekdays.

Thanks for your help.

**Comment by guilhemmarchand** [comment]
Hi @manderso7 

My apologises I didn't see a notification from Splunk Answers, I need to check why ;-)
Happy you like the trackMe promise, let's convert that in results!

I will review your questions very shortly, if you have some screenshots of the use case we are speaking about this would help understand it better.

Thanks

Guilhem

**Comment by manderso7** [comment]
No need to apologize, thanks for looking.
![2020-07-01 09_32_54-TrackMe _ Splunk](https://user-images.githubusercontent.com/9535390/86249967-f7bebc80-bb7d-11ea-8e07-7c5558969b6b.png)
![2020-07-01 09_32_38-TrackMe _ Splunk](https://user-images.githubusercontent.com/9535390/86249974-fa211680-bb7d-11ea-90ba-7a107b2630b1.png)

**Comment by guilhemmarchand** [comment]
@manderso7 
Thanks, I understand better the use case and the question.
 
1. 

_When I modify a data source, choosing "auto lagging" comes back with unexpected results. In particular, I have a data source that only ingests M-F, between 9am and 5pm. The auto percentile lag of this data source for 7-30 days at 1-3 seconds. How would I go about getting a longer average lag time for this source?_


**So effectively, TrackMe records 2 essential metrics can it come to lagging:**

- lag_ingestion_sec: delta between index time and latest event timestamp
- lag_event_sec: delta between now when the measure is taken, and the latest event timestamp

When you define a limit fo the max lagging, say 3600 seconds (1 hour), trackMe will define a green / red (essentially) status based on the condition if the max lagging is  superior to 300 seconds either for the event lag, or for the event lag, or for both.

**If you open of the single forms on the left of the screen you shared, you will see:**

eval delta=(indextime-_time)

This measure refers to the delta in seconds between the indexing time of the event (when it was indexed versus the event timestamp as parsed by Splunk)

So basically what it says is that your events were indexed with a very low latency.

The max lagging value you can interact is effectively applied against both measures, and any of these 2 measures would influence the state of it.

Normally on this specific data source, you should have one of the 2 measures showing the difference between and the latest time of you had events for it, can you confirm ? (it's not visible in the screenshot)

Technically you would define a max lagging value which matches the level where you want to get alerted, this would indeed affect your capability to detect ingestion latency on this data source in favor of detecting if you basically have events as expected or not. 

I need to think about more about this use case, basically a batch use use. (and find a way to allow dissociating both if that is required)

2. 

_In addition, how can I tell trackme to not show an alert state on the same sourcetype on a monday morning, since it hasn't gotten any events since the friday before? I don't want to set the lag time as too high as that will interfere w/ monitoring during weekdays._

The week days definition basically influence the results in the following way:

- If a data source would be red due to monitoring conditions, but if the current days is not a monitored day, the state will be orange instead of red, the status message would reflect the information, and the OOTB alert would not trigger.


Let me know your thinking / understanding of my answer ;-)

**Comment by manderso7** [comment]
1. Is this what you're looking for?
![2020-07-01 11_03_42-TrackMe _ Splunk](https://user-images.githubusercontent.com/9535390/86259901-994c0b00-bb8a-11ea-9ff1-1812f831761b.png)
To verify we're on the same page, you're talking about the auto lagging option measuring the longest time between events?
2. That makes sense. The status can show orange, until it turns green when data comes in, or red when it reaches the max data lag, ignoring the unmonitored days.

**Comment by guilhemmarchand** [comment]
@manderso7 

Thanks, yes this is correct.

The above shows that the time of your writing, there was 2393 seconds between now (when the last execution of the tracker exactly), and that the ingestion was near 0 seconds. (certainly between 0 and 0.5 but round at one decimal)


Yes I think we are on the same page, you are talking about:

![image](https://user-images.githubusercontent.com/7503059/86265547-d41a6700-bbbb-11ea-8cb0-545b4f1e47cd.png)

![image](https://user-images.githubusercontent.com/7503059/86265628-e98f9100-bbbb-11ea-93ab-82e61c536400.png)

What this feature does is basically calculating different versions of what you could choose for the max lagging value, instead of defining it yourself statically in the previous window.

Note that once defined, this will not change anymore and will not be re-calculated or updated to reflect the current behaviour of the ingestion latency)

So to be clear, this is the ingestion latency max accepted definition with a feature that provides different options you might want to choose.

Does it make sense ?

Guilhem

**Comment by manderso7** [comment]
Yes, thanks. Though I'm just wondering why the suggested lag times are so different from the lag times witnessed in the data source above. If the suggested lag times could be more similar to the visible gaps in the data they would be more useful.
![image](https://user-images.githubusercontent.com/9535390/86354425-087f3900-bc37-11ea-81f2-f5a0818723fb.png)

**Comment by guilhemmarchand** [comment]
@manderso7 

Because technically you didn't have any latency at the ingestion time (the difference between the Splunk _indextime and the _time which is extracted from the events according to your sourcetype definition or what Splunk used for the time stamp recognition)

The latency at ingestion time is what shows the windows exposing the 4 choices for an auto determinate potential choice for the max lagging.

The delta between now, and the latest time you had data (which in true is the latest time recognized by Splunk, the _time) is what you call the visible gaps in the data, and is called lagging event in the app.

Note: I understand there might some confusion in these single forms as they don't explicitly say it and I will fix that)

So basically, the gaps between 2 bunch of data is the value for lag event (and will grow by time until you get new data) and the lag ingestion is the latency at which your events were indexed in Splunk.

One option I could provide in TrackMe would be to allow you to choose if you want to consider the lag event or the lag ingestion for the calculation of the values, would you think this addresses the problem ?

**Comment by manderso7** [comment]
I think the calculation should be based on the amount of time between events, and not the latency of events being ingested.
Also, I think _indextime is more valuable than _time to measure lag ingestion. 
If possible, maybe have the option you suggested, and it's worth it to see if that shows the data more clearly.
Thanks for listening to my suggestions.

**Comment by guilhemmarchand** [comment]
More than happy to take your suggestions in consideration ;-)

- the amount of time between events is a third metric that can indeed be interested specially for batch like use cases, I will consider it
- _indextime is already taken in charge and consideration, when you calculate the latency you need to understand the time taken by an event to be indexed on the disk (the _indextime) and the time the event was produced (the _time), this is already at the root of everything and is the lag ingestion as explained previously

**Comment by guilhemmarchand** [comment]
@manderso7 

Work is being achieved to better handle a similar batch like use case, and shall be on-boarded within Version 1.2.18.

Guilhem

**Comment by guilhemmarchand** [comment]
Hi @manderso7 

TrackMe version 1.2.19 has been published with some new features that should address your use case.

Especially, you can now choose from which KPI the alert status will be derived, as well as a better rendering of the differences between the lag at ingestion time and the lag event.

The documentation: https://trackme.readthedocs.io

Has been seriously reviewed and updated too.

Let me know your thinking, if this addresses properly your requirements, in such a case we will close this issue.

Guilhem
